Add section about sympy.stats

rouckas opened the pull request:
It has become a bit longer than I expected and I don't even mention the sampling functionality. But I can shorten it if needed.

rouckas committed:
Add section about sympy.stats

asmeurer commented:
+1

rouckas committed:
Fix case of SymPy in stats.tex

Upabjojr commented on stats.tex:
```diff
@@ -0,0 +1,50 @@
+%% Stats
+The \verb|sympy.stats| module provides random variable types and methods for
+determination of statistical properties of expressions involving random
+variables. The variables can be either continuous or discontinuous, which are
```
continuous or discrete

rouckas replied:
Thanks for the catch, I'll fix it.

Upabjojr commented on stats.tex:
```diff
@@ -0,0 +1,50 @@
+%% Stats
+The \verb|sympy.stats| module provides random variable types and methods for
+determination of statistical properties of expressions involving random
```
_statistical properties_ is a bit ambiguous. What _sympy.stats_ currently does, is to generate the symbolic integrals/summations for statistical expressions. No properties are determined, symbolic calculations are performed.

rouckas replied:
But the integrals/sums are automatically evaluated. From the end user point of view, E(X) returns the expectation, etc... I would like to leave it as it is for the sake of simplicity, If you agree.

Upabjojr replied:
Being a technical paper, I suggest to mention some internal stuff. Most of the algorithms in _sympy.stats_ are about generating integrals/summations. They can be shown with `evaluate=False`.

The term _properties_ is too general, it could mean anything. What about _symbolic values_ instead?

rouckas replied:
I will add this information about evaluation

Upabjojr replied:
`for determination` or `for the determination`? which one is better?

rouckas replied:
I prefer `for determination` but I'm not a native speaker

asmeurer replied:
I believe "for the determination" is more grammatically correct, but both sound awkward. "for computing" sounds better.

Upabjojr commented on stats.tex:
```diff
@@ -0,0 +1,50 @@
+\texttt{Binomial}, \texttt{Bernoulli}, and many others.
+
+Properties of random expressions can be calculated using, e.g.,
+\texttt{E} and \texttt{variance} to calculate expectation and variance.
```
`\texttt{expectation} (abbreviated \texttt{E})`

rouckas replied:
OK, I'll change it.

Upabjojr commented on stats.tex:
```diff
@@ -0,0 +1,50 @@
+distributions, including \texttt{Uniform}, \texttt{Poisson}, \texttt{Normal},
+\texttt{Binomial}, \texttt{Bernoulli}, and many others.
+
+Properties of random expressions can be calculated using, e.g.,
```
Expectation, variance, and covariance map random expressions to real numbers. I wouldn't call that properties.

rouckas replied:
I think that a real number assigned to a mathematical object can be called a property. [Wolfram](https://reference.wolfram.com/language/guide/RandomVariables.html) also calls it properties:

> In the Wolfram Language, you can directly compute several dozen properties from symbolic distributions

Upabjojr commented on stats.tex:
```diff
@@ -0,0 +1,50 @@
+\texttt{E} and \texttt{variance} to calculate expectation and variance.
+Conditions on random variables can be defined with inequalities, equalities,
+and logical operators and their overall probabilities are obtained using
+\texttt{P}. The features can be illustrated on a model of two dice throws:
```
probability( ) abbreviated P( ). I would like to change _P_ in future versions of SymPy to return an unevaluated probability expression.

rouckas replied:
OK

rouckas replied:
However, probability is not exposed when importing from sympy.stats:

```
>>> from sympy.stats import probability
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'probability'

```

One would need to import it from `sympy.stats.rv`

Upabjojr commented on stats.tex:
```diff
@@ -0,0 +1,50 @@
+\end{verbatim}
+
+More information on the \texttt{sympy.stats} module can be found in
+\cite{StatsMRocklin}.
```
Are you sure it's the right way to render a citation?

Upabjojr replied:
@mrocklin are there other papers/links concerning _sympy.stats_ ?

scolobb replied:
To me, citing a resource in this way looks pretty natural.

Upabjojr replied:
I'm talking about the rendering style.

rouckas replied:
I know about the two papers referenced in #16 

You mean saying "in \cite{StatsMRocklin}"? I think it depends on the final format of refences. But I agree that it may need to be changed, perhaps to "in Ref. \cite..." or "in paper by \cite..." in case of author-year format.

scolobb replied:
@Upabjojr could you please explain what you mean exactly by rendering style?

In the version I'm looking at I see "can be found in [36]".  That's exactly how I saw citations rendered in very many articles and it looks absolutely natural to me (except that I'd probably add a non-breaking space `~` between "in" and `\cite`.

@rouckas I've never seen "in Ref. \cite" and I find it weird (but then YMMV).  "In the paper by \cite" is even weirder to me, because [36] is a paper, and not a list of authors.  "In the paper \cite" is something see just as often as "in \cite".

rouckas replied:
@scolobb For example in Journal of Chemical Physics, you could write "in Ref. 11" or "by Esaulov et al.^{11}". I see that peerj template uses author-year citations, so we may be able to write something like "For more information see Rocklin and Terrel (2012)" or "information can be found in (Rocklin and Terrel, 2012)"

asmeurer replied:
Maybe he means to use `in~\cite{StatsMRocklin}`. I'm unclear if `~` is supposed to be used when it's not a parenthetical.

scolobb replied:
@rouckas All right, I see your point.  Indeed, the citation style varies in different domains.  AFAIK, in maths and computer science the form "in [x]" is quite widespread and accepted.  On the other hand, the way citations are typeset is given by the bibliography style (a LaTeX style file), so I'd go with the following strategy: if the target journal has a recommended bibliography style, we take; if not, we go with what we have now to avoid the extra effort of re-typesetting all the citations :-)

@asmeurer I'd recommend using "in~\cite{...}" because it prevents line breaks between the "in" and the citation number, which (I think) is considered ugly.

Upabjojr commented on stats.tex:
```diff
@@ -0,0 +1,50 @@
+The \verb|sympy.stats| module provides random variable types and methods for
+determination of statistical properties of expressions involving random
+variables. The variables can be either continuous or discontinuous, which are
+further divided into finite and infinite. They are associated
```
the latter being further divided

rouckas replied:
OK

Upabjojr commented on stats.tex:
```diff
@@ -0,0 +1,50 @@
+\begin{verbatim}
+>>> from sympy.stats import Die, P, E
+>>> X, Y = Die("X"), Die("Y")
+>>> P(Eq(X, 6) & Eq(Y, 6))
```
Add example with `evaluate=False`, so that it is clear how it works.

rouckas replied:
OK

rouckas replied:
Evaluate=False does not work with finite RV for me:

```
>>> E(Die("X"), evaluate=False)
7/2
```

Upabjojr commented on stats.tex:
```diff
@@ -0,0 +1,50 @@
+>>> from sympy.stats import Maxwell, density
+>>> kT, m, x = symbols("kT m x", positive=True)
+>>> v = Maxwell("v", sqrt(kT/m))
+>>> E(v)                                # mean velocity
```
Example with `evaluate=False` would be nice here.

rouckas replied:
OK

rouckas committed:
Revision of stats.tex

Based on comments in PR #187 I have added a bit more information
about the internals and fixed some minor errors.

rouckas commented:
@Upabjojr what do you think about the updated text?

Upabjojr commented on stats.tex:
```diff
@@ -0,0 +1,57 @@
+%% Stats
+The \verb|sympy.stats| module provides random variable types and methods for
+determination of statistical properties of expressions involving random
+variables. The variables can be either continuous or discrete, the latter being
```
_variables, which can be either_

Upabjojr replied:
_the latter ones_

rouckas replied:
OK, I'll change it to:

```
The \verb|sympy.stats| module provides random variable types and methods for
determination of statistical properties of expressions involving random
variables, which can be either continuous or discrete, the latter ones being
further divided into finite and infinite. The variables are associated
with probability densities on corresponding domains and internally defined
in terms of probability spaces. 
```

(changing also "They" to "The variables" in the second sentence, because it wouldn't be clear what "They" refers to)

rouckas committed:
Revision 2 of stats.tex

More minor corrections

asmeurer commented:
@rouckas are all the comments addressed?

rouckas commented:
@asmeurer I think so. @Upabjojr do you agree?

Upabjojr commented:
I think this PR is OK.

I would like future versions of SymPy to have _E_ and _P_ return unevaluated expressions whereas the evaluated expression would be returned by _expectation_ and _probability_ (as _Integral_, _Derivative_, _Limit_ vs _integrate_, _derive_, _limit_). Anyway, this is an issue for SymPy.
